### PR TITLE
DistributedPubSub: Fixed race-condition issue when handling Topic/Group actors

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
Fixed https://github.com/akkadotnet/akka.net/issues/2055

@jordansjones could you help me to debug this change? I missed somethings and specs are failing
You can find an original scala code in the linked issue

### CODE REVIEW

**Comment by JordanJones** [comment]
@alexvaluyskiy, I'll take a look into this tonight or tomorrow sometime :)

**Comment by alexvaluyskiy** [comment]
@JordanJones I've rebased this PR. So it contains (should) the same code, as we have in Akka Jvm 2.4.8. So, if you find any difference, please tell us

**Comment by jordansjones** [comment]
Having spent most of last night and part of today looking this over, the only thing that looks out of place to me is the `HandlePrune` method.

```csharp
        private void HandlePrune()
        {
            foreach (var entry in _registry)
            {
                var owner = entry.Key;
                var bucket = entry.Value;

                var oldRemoved = bucket.Content
                    .Where(kv => (bucket.Version - kv.Value.Version) > _settings.RemovedTimeToLive.TotalMilliseconds)
                    .Select(kv => kv.Key);

                if (oldRemoved.Any())
                {
                    _registry.Add(owner, new Bucket(bucket.Owner, bucket.Version, bucket.Content.RemoveRange(oldRemoved)));
                }
            }
        }
```

It seems like it should look like this (minor change):
```csharp
        private void HandlePrune()
        {
            foreach (var entry in _registry)
            {
                var owner = entry.Key;
                var bucket = entry.Value;

                var oldRemoved = bucket.Content
                    .Where(kv => (bucket.Version - kv.Value.Version) > _settings.RemovedTimeToLive.TotalMilliseconds)
                    .Select(kv => kv.Key)
                    .ToArray();

                if (oldRemoved.Any())
                {
                    _registry[owner] = new Bucket(bucket.Owner, bucket.Version, bucket.Content.RemoveRange(oldRemoved));
                }
            }
        }
```

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs [hunk 1]
```diff
@@ -183,12 +186,11 @@ public DistributedPubSubMediator(DistributedPubSubSettings settings)
             });
             Receive<Publish>(publish =>
             {
-                var topic = Uri.EscapeDataString(publish.Topic);
-                var path = Self.Path / topic;
+                string path = Internal.Utils.MkKey(Self.Path / Internal.Utils.EncodeName(publish.Topic));
```
Comment (Horusiath) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs]: Please use full naming convention. We rarely use abbreviations in C#: therefore `MkKey` &rArr; `MakeKey`.

**Comment by alexvaluyskiy** [comment]
@jordansjones I've found this one
```csharp
        private void Remove(IActorRef actorRef)
        {
            Subscribers.Remove(actorRef);
            if (Subscribers.Count == 0 && !Context.GetChildren().Any())
            {
                PruneDeadline = Deadline.Now + EmptyTimeToLive;
            }
        }
```
should be
```csharp
        private void Remove(IActorRef actorRef)
        {
            if (Subscribers.Contains(actorRef))
            {
                Subscribers.Remove(actorRef);
            }
            if (Subscribers.Count == 0 && !Context.GetChildren().Any())
            {
                PruneDeadline = Deadline.Now + EmptyTimeToLive;
            }
        }
```

**Comment by Horusiath** [comment]
@alexvaluyskiy isn't that a bit redundant? `if (Subscribers.Contains(actorRef)) Subscribers.Remove(actorRef);`

**Comment by alexvaluyskiy** [comment]
Yes. but this latest change don't work. And I'm trying to figure out why

**Comment by alexvaluyskiy** [comment]
Still fails
```
[Node1][FAIL] Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorNode1.DistributedPubSubMediatorSpecs
[Node1][FAIL-EXCEPTION] Type: Xunit.Sdk.EqualException
--> [Node1][FAIL-EXCEPTION] Message: Assert.Equal() Failure
Expected: 4
Actual:   5
--> [Node1][FAIL-EXCEPTION] StackTrace:    at Xunit.Assert.Equal[T](T expected, T actual, IEqualityComparer`1 comparer) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.assert\Asserts\EqualityAsserts.cs:line 35
   at Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorSpec.<>c__DisplayClass24_0.<AwaitCount>b__0() in C:\Projects\akka.net\src\contrib\cluster\Akka.Cluster.Tools.Tests.MultiNode\PublishSubscribe\DistributedPubSubMediatorSpec.cs:line 289
   at Akka.TestKit.TestKitBase.AwaitAssert(Action assertion, Nullable`1 duration, Nullable`1 interval) in C:\Projects\akka.net\src\core\Akka.TestKit\TestKitBase_AwaitAssert.cs:line 48
   at Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorSpec.AwaitCount(Int32 expected) in C:\Projects\akka.net\src\contrib\cluster\Akka.Cluster.Tools.Tests.MultiNode\PublishSubscribe\DistributedPubSubMediatorSpec.cs:line 286
   at Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorSpec.<DistributedPubSubMediator_must_remove_terminated_users>b__30_0() in C:\Projects\akka.net\src\contrib\cluster\Akka.Cluster.Tools.Tests.MultiNode\PublishSubscribe\DistributedPubSubMediatorSpec.cs:line 443
   at Akka.TestKit.TestKitBase.<>c__DisplayClass116_0.<Within>b__0() in C:\Projects\akka.net\src\core\Akka.TestKit\TestKitBase_Within.cs:line 38
   at Akka.TestKit.TestKitBase.Within[T](TimeSpan min, TimeSpan max, Func`1 function, String hint, Nullable`1 epsilonValue) in C:\Projects\akka.net\src\core\Akka.TestKit\TestKitBase_Within.cs:line 79
   at Akka.TestKit.TestKitBase.Within(TimeSpan min, TimeSpan max, Action action, String hint, Nullable`1 epsilonValue) in C:\Projects\akka.net\src\core\Akka.TestKit\TestKitBase_Within.cs:line 38
   at Akka.TestKit.TestKitBase.Within(TimeSpan max, Action action, Nullable`1 epsilonValue) in C:\Projects\akka.net\src\core\Akka.TestKit\TestKitBase_Within.cs:line 26
   at Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorSpec.DistributedPubSubMediator_must_remove_terminated_users() in C:\Projects\akka.net\src\contrib\cluster\Akka.Cluster.Tools.Tests.MultiNode\PublishSubscribe\DistributedPubSubMediatorSpec.cs:line 436
   at Akka.Cluster.Tools.Tests.MultiNode.PublishSubscribe.DistributedPubSubMediatorSpec.DistributedPubSubMediatorSpecs() in C:\Projects\akka.net\src\contrib\cluster\Akka.Cluster.Tools.Tests.MultiNode\PublishSubscribe\DistributedPubSubMediatorSpec.cs:line 302
```

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/PerGroupingBuffer.cs [hunk 2]
```diff
@@ -31,12 +31,12 @@ public void BufferOr(string grouping, object message, IActorRef originalSender,
             }
         }
 
-        public void RecreateAndForwardMessagesIfNeeded(string grouping, IActorRef recipient)
+        public void RecreateAndForwardMessagesIfNeeded(string grouping, Func<IActorRef> recipient)
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/PerGroupingBuffer.cs]: I've spend many and many hours to fix this parameter :( A long debug story

**Comment by alexvaluyskiy** [comment]
@akkadotnet/core ready to review. The latest my PR for PubSub

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/Topics.cs [hunk 3]
```diff
@@ -195,5 +237,20 @@ public static object WrapIfNeeded(object message)
...
         }
+
+        public static string MakeKey(IActorRef actorRef)
+        {
+            return MakeKey(actorRef.Path);
+        }
+
+        public static string EncodeName(string name)
+        {
+            return name == null ? null : Uri.EscapeDataString(name);
```
Comment (rogeralsing) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/Topics.cs]: Is `null` really a valid value here?

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs [hunk 4]
```diff
@@ -183,12 +186,11 @@ public DistributedPubSubMediator(DistributedPubSubSettings settings)
             });
             Receive<Publish>(publish =>
             {
-                var topic = Uri.EscapeDataString(publish.Topic);
-                var path = Self.Path / topic;
+                string path = Internal.Utils.MakeKey(Self.Path / Internal.Utils.EncodeName(publish.Topic));
```
Comment (rogeralsing) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs]: I would probably have used a `using static Internal.Utils` and just have `MakeKey` and `EnodeName` in the code, as it is easier on the eye and the usage varies between code inside and outside of the `Internal` namespace.

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/Topics.cs (hunk 3 above)
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/Topics.cs]: Scala version doesn't have this check
https://github.com/akka/akka/blob/master/akka-cluster-tools/src/main/scala/akka/cluster/pubsub/DistributedPubSubMediator.scala#L281

**Comment by rogeralsing** [comment]
I'm missing context as I've been off for a while, but I find it easy to read and I see no obvious problems

**Comment by alexvaluyskiy** [comment]
@rogeralsing we have at least 90% code coverage in PubSub extension. So, this code should work fine. You could merge it :)

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs (hunk 4 above)
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/DistributedPubSubMediator.cs]: I'm planning to do changes like you mentioned, right before Cluster.Tools become stable

**Pull request merged**